src/python/pants/backend/python/lint/yapf/rules.py

```diff
@@ -0,0 +1,155 @@
+    check_only: bool
+    inplace: bool
+
+
+@dataclass(frozen=True)
+class Setup:
+    process: Process
+    original_digest: Digest
+
+
+def generate_argv(
+    source_files: SourceFiles, yapf: Yapf, check_only: bool, inplace: bool
+) -> Tuple[str, ...]:
+    args = [*yapf.args]
+    if check_only:
+        # If "--diff" is passed, yapf returns zero when no changes were necessary and
+        # non-zero otherwise
+        args.append("--diff")
+    if inplace:
+        # The "--in-place" flag makes yapf to actually reformat files
+        args.append("--in-place")
+    if yapf.config:
+        args.extend(["--style", yapf.config])
+    args.extend(source_files.files)
+    return tuple(args)
+
+
+@rule(level=LogLevel.DEBUG)
+async def setup_yapf(setup_request: SetupRequest, yapf: Yapf) -> Setup:
+    yapf_pex_get = Get(
+        VenvPex,
+        PexRequest(
+            output_filename="yapf.pex",
+            internal_only=True,
+            requirements=PexRequirements(yapf.all_requirements),
+            interpreter_constraints=InterpreterConstraints(yapf.interpreter_constraints),
+            main=yapf.main,
+        ),
+    )
+    source_files_get = Get(
+        SourceFiles,
+        SourceFilesRequest(field_set.sources for field_set in setup_request.request.field_sets),
+    )
+    source_files, yapf_pex = await MultiGet(source_files_get, yapf_pex_get)
+
+    source_files_snapshot = (
+        source_files.snapshot
+        if setup_request.request.prior_formatter_result is None
+        else setup_request.request.prior_formatter_result
+    )
+
+    config_files = await Get(
+        ConfigFiles, ConfigFilesRequest, yapf.config_request(source_files_snapshot.dirs)
+    )
+
+    input_digest = await Get(
+        Digest, MergeDigests((source_files_snapshot.digest, config_files.snapshot.digest))
+    )
+
+    process = await Get(
+        Process,
+        VenvPexProcess(
+            yapf_pex,
+            argv=generate_argv(
+                source_files,
+                yapf,
+                check_only=setup_request.check_only,
+                inplace=setup_request.inplace,
+            ),
+            input_digest=input_digest,
+            output_files=source_files_snapshot.files,
+            description=f"Run yapf on {pluralize(len(setup_request.request.field_sets), 'file')}.",
+            level=LogLevel.DEBUG,
+        ),
+    )
+    return Setup(process, original_digest=source_files_snapshot.digest)
+
+
+@rule(desc="Format with yapf", level=LogLevel.DEBUG)
+async def yapf_fmt(request: YapfRequest, yapf: Yapf) -> FmtResult:
+    if yapf.skip:
+        return FmtResult.skip(formatter_name="yapf")
+    setup = await Get(Setup, SetupRequest(request, check_only=False, inplace=True))
+    result = await Get(ProcessResult, Process, setup.process)
+    return FmtResult.from_process_result(
+        result,
+        original_digest=setup.original_digest,
+        formatter_name="yapf",
+        strip_chroot_path=True,
```

Comment:
Is this necessary for yafp? Some tools output relative paths to begin with.

Reply:
No, I don't think so. I've removed these.